roachtest: tpccbench/nodes=9/cpu=4/chaos/partition failed

SHA: https://github.com/cockroachdb/cockroach/commits/2859f9b35f1ed6d1a4e960a106e951d81de21287

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=tpccbench/nodes=9/cpu=4/chaos/partition PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1535775&tab=artifacts#/tpccbench/nodes=9/cpu=4/chaos/partition

```
The test failed on branch=release-19.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20191012-1535775/tpccbench/nodes=9/cpu=4/chaos/partition/run_1
	test_runner.go:704: test timed out (10h0m0s)

```

[(roachtest).tpccbench/nodes=9/cpu=4/chaos/partition failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1676715&tab=buildLog) on [master@52261c39c65aecf960e3c8d54089a66d378af610](https://github.com/cockroachdb/cockroach/commits/52261c39c65aecf960e3c8d54089a66d378af610):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200107-1676715/tpccbench/nodes=9/cpu=4/chaos/partition/run_1
	cluster.go:2189,tpcc.go:720,tpcc.go:561,test_runner.go:716: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2180
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2188
		    main.runTPCCBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:720
		    main.registerTPCCBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:561
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:716
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 7: dead
```

<details><summary>Repro</summary><p>


Artifacts: [/tpccbench/nodes=9/cpu=4/chaos/partition](https://teamcity.cockroachdb.com/viewLog.html?buildId=1676715&tab=artifacts#/tpccbench/nodes=9/cpu=4/chaos/partition)
```
make stressrace TESTS=tpccbench/nodes=9/cpu=4/chaos/partition PKG=./pkg/roachtest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpccbench/nodes=9/cpu=4/chaos/partition failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1693614&tab=buildLog) on [master@e2c2be6cbef07c7f2ee6341816f81f72a206c32a](https://github.com/cockroachdb/cockroach/commits/e2c2be6cbef07c7f2ee6341816f81f72a206c32a):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200116-1693614/tpccbench/nodes=9/cpu=4/chaos/partition/run_1
	cluster.go:2201,tpcc.go:720,tpcc.go:561,test_runner.go:716: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2192
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2200
		    main.runTPCCBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:720
		    main.registerTPCCBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:561
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:716
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 2: dead
```

<details><summary>Repro</summary><p>


Artifacts: [/tpccbench/nodes=9/cpu=4/chaos/partition](https://teamcity.cockroachdb.com/viewLog.html?buildId=1693614&tab=artifacts#/tpccbench/nodes=9/cpu=4/chaos/partition)
```
make stressrace TESTS=tpccbench/nodes=9/cpu=4/chaos/partition PKG=./pkg/roachtest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpccbench/nodes=9/cpu=4/chaos/partition failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1705428&tab=buildLog) on [master@545c96573d36519c27115278187fdd441f54a64d](https://github.com/cockroachdb/cockroach/commits/545c96573d36519c27115278187fdd441f54a64d):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200123-1705428/tpccbench/nodes=9/cpu=4/chaos/partition/run_1
	cluster.go:2201,tpcc.go:720,tpcc.go:561,test_runner.go:719: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2192
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2200
		    main.runTPCCBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:720
		    main.registerTPCCBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpcc.go:561
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:719
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - unexpected node event: 7: dead

```

<details><summary>More</summary><p>


Artifacts: [/tpccbench/nodes=9/cpu=4/chaos/partition](https://teamcity.cockroachdb.com/viewLog.html?buildId=1705428&tab=artifacts#/tpccbench/nodes=9/cpu=4/chaos/partition)
Related:
- #44287 roachtest: tpccbench/nodes=9/cpu=4/chaos/partition failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44259 roachtest: tpccbench/nodes=9/cpu=4/chaos/partition failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-provisional_202001221958_v2.1.11](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-provisional_202001221958_v2.1.11) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpccbench%2Fnodes%3D9%2Fcpu%3D4%2Fchaos%2Fpartition.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/logictestccl/testdata/logic_test/zone
+++ b/pkg/ccl/logictestccl/testdata/logic_test/zone
@@ -478,6 +478,14 @@ SELECT zone_name, config_sql FROM [SHOW ZONE CONFIGURATIONS] WHERE zone_name = '
 test.parent_modify  ALTER TABLE test.public.parent_modify CONFIGURE ZONE USING
                     gc.ttlseconds = 700
 
+# Regression test for #44231.
+statement ok
+CREATE TABLE t44231 (x INT, INDEX i (x));
+ALTER INDEX t44231@i CONFIGURE ZONE USING num_replicas=COPY FROM PARENT;
+ALTER INDEX t44231@i CONFIGURE ZONE USING num_replicas=COPY FROM PARENT;
+ALTER INDEX t44231@i CONFIGURE ZONE USING num_replicas=COPY FROM PARENT
+
+
 # ------------------------------------------------------------------------------
 # Regression test for #36348; place this at the bottom of this file.
 # ------------------------------------------------------------------------------

--- a/pkg/config/zone.go
+++ b/pkg/config/zone.go
@@ -504,7 +504,8 @@ func (z *ZoneConfig) Validate() error {
 
 // InheritFromParent hydrates a zones missing fields from its parent.
 func (z *ZoneConfig) InheritFromParent(parent *ZoneConfig) {
-	if z.NumReplicas == nil {
+	// Allow for subzonePlaceholders to inherit fields from parents if needed.
+	if z.NumReplicas == nil || (z.NumReplicas != nil && *z.NumReplicas == 0) {
 		if parent.NumReplicas != nil {
 			z.NumReplicas = proto.Int32(*parent.NumReplicas)
 		}

